A rudimentary Nim bundle

shalabhc opened the pull request.

kirbyfan64 commented:
YEAH!!!

kirbyfan64 commented:
(I was going to do this, but you beat me to it!)

kirbyfan64 commented:
I don't think this properly supports Nim's case insensitivity. Technically, these are all equivalent:

```nim
proc
p_ROC
p_R_O_C
P__________________________________r_O_c
```

shalabhc commented:
Oh.. I didn't realize this applies to keywords too. Will fix.

shalabhc commented:
@kirbyfan64 - the case insensitivity for keywords is now fixed. It appears the first character is case sensitive and more than one contiguous underscore is not allowed. I've also fixed and simplified the number parsing. Please take another look when you get a chance!

kirbyfan64 commented:
I haven't actually tried this, but it looks great!!

One more thing: the Nim style guide (and 99% of Nim programmers) capitalize the first letter of a type. Should it be highlighted in a special way like the MoonScript bundle?

An interesting extension (likely for a later time) would be to use [nimsuggest](http://nim-lang.org/docs/nimsuggest.html) to implement [`show-doc-at-cursor`](http://howl.io/doc/manual/editing.html#documentation-popup) support. I think this is good enough for now, though.

If (when?) this gets merged in, I'll add Howl to the list of editors with Nim support.

nilnor commented on bundles/nim/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## License
+
+Copyright 2012-2015 The Howl Developers
```
Should probably be just 'Copyright 2015' to avoid mental time warps.

shalabhc commented:
@kirbyfan64 - great suggestions.

> the Nim style guide (and 99% of Nim programmers) capitalize the first letter of a type. Should it be highlighted in a special way like the MoonScript bundle?

Yes, this sounds reasonable. Perhaps with nimsuggest it may even be possible to accurately determine if an identifier is a type, but integrating that with lexing doesn't seem simple.

> An interesting extension (likely for a later time) would be to use nimsuggest to implement show-doc-at-cursor support.

Yes this sounds like a great idea (as a separate enhancement in the future). In fact we could also nimsuggest for completions. :)

nilnor commented on bundles/nim/init.moon:
```diff
@@ -0,0 +1,33 @@
+unload = -> howl.mode.unregister 'nim'
+
+howl.config.define
+  name: 'nim_executable',
```
Considering the suggestion by @kirbyfan64 to incorporate `nimsuggest` at a later stage, I think it's a good idea to contemplate having a `nim_bin_dir` variable or something similar instead. This could be used to find the `nim` executable as well as other related executables. Should default to `/usr/bin` as that's where distribution package would typically place the files, which would make it "just work" at a later stage.

kirbyfan64 replied:
@nilnor But `nimsuggest` is usually installed somewhere into `~/.nimble`...

nilnor replied:
@kirbyfan64 
> But nimsuggest is usually installed somewhere into ~/.nimble...

Oh? The page for nimsuggest you linked to says it doesn't install properly via nimble and suggests compiling it and putting it in the same bin path as nim, which made me believe that was the standard way of doing it.

@shalabhc Even if we keep this variable it could default to just `nim`, right? Provided that `nim` is in the path somewhere already it would still just work then I think.

nilnor replied:
Or as an alternative we could auto-find it given the current $PATH if we want the full path of the binary.

shalabhc replied:
Using 'nim' as the default is a good idea - it should just work if it's in the path. We don't need to know the full path.

Re nimsuggest, currently I've copied it in the directory containing nim, but I believe when fixed it will install automatically via nimble and may not be in the same directory. So that might need its own variable - not sure what we'd call that - 'nimble_executable' or 'nim_nimble_executable' (in case we want to namespace everything by the bundle name.

nilnor commented:
(will continue reviewing later)

kirbyfan64 commented:
After actually trying this, I noticed none of the built-in types are highlighted (e.g. `string`, `int`, etc.), nor are any of the "fake types" (like `varargs`). Is this intentional?

shalabhc commented:
@kirbyfan64 
> none of the built-in types are highlighted

Ah yes this should be fixed. The primary reason is  laziness on my part, plus not having an exhaustive list of these builtins. Any idea how I can get one?

kirbyfan64 commented:
http://nim-lang.org/docs/manual.html#types isn't quite exhaustive, but it's close.

nilnor commented on bundles/nim/nim_lexer.moon:
```diff
@@ -0,0 +1,94 @@
+
```
Missing copyright notices at the top.

nilnor commented on bundles/nim/nim_mode.moon:
```diff
@@ -0,0 +1,61 @@
+  indentation: {
+    more_after: {
+      '[([{:=]%s*$'
+      r'^\\s*\\b(const|enum|let|type|proc|func|iterator|macro|template|method|object|tuple|var)\\s*$',
```
The `\\b` seems excessive since the pattern is already anchored.

nilnor commented:
What I really miss for this bundle is a syntax example file.

nilnor commented on bundles/nim/nim_lexer.moon:
```diff
@@ -0,0 +1,94 @@
+  number += c('number', number_with_point) * (exponent_suffix^-1) * (float_size_suffix^-1)
+  number += (integer * integer_size_suffix^-1)
+  number *= #-(alpha + digit + S'_') -- no alphanum should be attached to the number
+  -- number = integer
```
Old code left in?

nilnor commented:
Finished looking it over, looks good as far as I can tell :) 

Like I commented above, a syntax example file containing all the lexed constructs is what I really miss, doing any kind of larger changes to the lexer later without breaking stuff will be unnecessarily hard later. Such a file also gives a very nice overview over exactly what is lexed and how it appears.

shalabhc commented:
I believe I've addressed all issues. In particular, anything that looks like a type (starts with uppercase) is styled as a class and all builtin system types are also styled as classes. I've also added an example file and made some bug fixes.

nilnor commented on bundles/nim/init.moon:
```diff
@@ -0,0 +1,36 @@
+      error "No file associated with current buffer."
+
+    exe = howl.app.editor.buffer.config.nim_executable or 'nim'
+    unless howl.io.File(exe).exists
```
This won't work when exe is "nim". Either forego the check completely or add a path search for finding the full path. It's a shame to lose the informational warning/error pointing to the correct configuration variable though if we just go with skipping the check.

shalabhc replied:
Good catch. Now fixed - please look when you get a chance.

nilnor commented:
All right, looks good!
